contracts/Keyboards.sol

```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.2;
+
+contract Keyboards {
+  enum KeyboardKind { 
+    SixtyPercent, 
+    SeventyFivePercent, 
+    EightyPercent, 
+    Iso105 
+  }
+
+  struct Keyboard {
+    KeyboardKind kind; 
+    // ABS = false, PBT = true
+    bool isPBT;
+    // tailwind filters to layer over
+    string filter;
+    // user who created it! 
+    address owner;
+  }
+
+  Keyboard[] public createdKeyboards;
+
+  event KeyboardCreated(
+    Keyboard keyboard
+  );
+
+  function create(
+    KeyboardKind _kind,
+    bool _isPBT,
+    string calldata _filter
+  ) external {
+    Keyboard memory newKeyboard = Keyboard({
+      kind: _kind,
+      isPBT: _isPBT,
+      filter: _filter,
+      owner: msg.sender
+    });
+
+    createdKeyboards.push(newKeyboard);
+    emit KeyboardCreated(newKeyboard);
+  }
+
+  function tip(address payable _owner) external payable  {
+    _owner.transfer(msg.value);
+  }
+
+  function getKeyboards() view public returns (Keyboard[] memory) {
```

Comment:
What happens if we want to modify this? say we want to add some sort of filter option to get keyboards for a specific owner or the list gets really long and we want to paginate it. Is there a way to make those updates?

Reply:
This is one of the fiddly things with Eth development. The contract is immutable, so if we want to modify this function we have to deploy a new contract, and that new version won't have the stored data from the previous one.

There are workarounds for that where you basically have a contract that just stores your data and then another contract that acts as the API around it, and the API contract can call the storage one and then the API one can be updated freely since it's stateless. That might make a good future tutorial for us, it's definitely something I'd like to learn more about! But I think for this tutorial we probably don't want to get into that in any detail.